framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js

```diff
@@ -42,47 +43,199 @@ class BlockSynchronizationMechanism {
+				),
+			);
+		}
+
+		try {
+			await deleteBlocksAfterHeight(
+				this.processorModule,
+				this.blocks,
+				lastCommonBlock.height,
 			);
-			const notInDifferentChain = ForkChoiceRule.isDifferentChain(
-				this.blocks.lastBlock,
-				networkLastBlock,
+		} catch (e) {
+			this.logger.debug(
+				`Failure deleting blocks after height ${
+					lastCommonBlock.height
+				} while reverting chain to last common block`,
 			);
+			throw e;
+		}
+	}
+
+	/**
+	 * Requests the last common block in common with the targeted peer.
+	 * In order to do that, sends a set of network calls which include a set of block ids
+	 * corresponding to the first block of descendent consecutive rounds (starting from the last one)
+	 * @param peer - The peer to target.
+	 * @return {Promise<Object | undefined>}
+	 * @private
+	 */
+	async _requestLastCommonBlock(peer) {
+		const [lastBlock] = await this.storage.entities.Block.get({
+			sort: 'height:desc',
+			limit: 1,
+		});
 
-			if (!validBlock || !notInDifferentChain) {
-				this.channel.invoke('network:applyPenalty', {
-					peerId: bestPeer.id,
-					penalty: 100,
-				});
-				this.channel.publish('chain:processor:sync', { block: receivedBlock });
-				throw err;
+		const blocksPerRequestLimit = 10; // Maximum number of block IDs to be included in a single request
+		const requestLimit = 10; // Maximum number of requests to be made to the remote peer
+
+		let numberOfRequests = 0; // Keeps track of the number of requests made to the remote peer
+		let numberOfBlocks = 0; // Keeps track of the number of block IDs to be included in a request
+		let currentRound = Math.floor(
+			lastBlock.height / this.constants.activeDelegates,
+		); // Keeps track of the round number it is  being used to compute the first block of the round to be included in the request payload
+		let highestCommonBlock; // Holds the common block returned by the peer if found.
+
+		while (
+			!highestCommonBlock &&
+			numberOfRequests < requestLimit &&
+			currentRound * this.constants.activeDelegates < this.bft.finalizedHeight
+		) {
+			const blockHeights = [];
+
+			while (
+				numberOfBlocks < blocksPerRequestLimit &&
+				currentRound * this.constants.activeDelegates < this.bft.finalizedHeight
+			) {
+				let height = currentRound * this.constants.activeDelegates;
+				if (height <= this.bft.finalizedHeight) {
+					// if the calculated height is smaller than finalized height we push finalized height
+					// on the array and currentRound -= 1 will stop the loop.
+					// The reason for this is stated in step 3.b of LIP-0014 under Block Synchronization Mechanism
+					height = this.bft.finalizedHeight;
+				}
+
+				blockHeights.push(height);
+				currentRound -= 1;
+				numberOfBlocks += 1;
 			}
-		} finally {
-			this.active = false;
+
+			const blocks = await this.storage.entities.Block.get(
+				{ height_in: blockHeights },
+				{
+					sort: 'height:asc',
+				},
+			);
+
+			const blockIds = blocks.map(block => block.id);
```

Comment:
didn't we have this logic in the blocks module? and i think it should be in the blocks module

Reply:
Moved inner while loop to a separate function within same file as it is specific logic to BSM